contracts/staking/aave/GoodAaveStaking.sol

```diff
@@ -0,0 +1,249 @@
+	uint32 public collectInterestGasCost = 250000;
+
+	/**
+	 * @param _token Token to swap DEFI token
+	 * @param _lendingPool LendingPool address
+	 * @param _ns Address of the NameService
+	 * @param _tokenName Name of the staking token which will be provided to staker for their staking share
+	 * @param _tokenSymbol Symbol of the staking token which will be provided to staker for their staking share
+	 * @param _tokenSymbol Determines blocks to pass for 1x Multiplier
+	 * @param _tokenUsdOracle address of the TOKEN/USD oracle
+	 * @param _incentiveController Aave incentive controller which provides AAVE rewards
+	 * @param _aaveUSDOracle address of the AAVE/USD oracle
+	 */
+	function init(
+		address _token,
+		address _lendingPool,
+		INameService _ns,
+		string memory _tokenName,
+		string memory _tokenSymbol,
+		uint64 _maxRewardThreshold,
+		address _tokenUsdOracle,
+		IAaveIncentivesController _incentiveController,
+		address _aaveUSDOracle
+	) public {
+		lendingPool = ILendingPool(_lendingPool);
+		DataTypes.ReserveData memory reserve =
+			lendingPool.getReserveData(_token);
+		initialize(
+			_token,
+			reserve.aTokenAddress,
+			_ns,
+			_tokenName,
+			_tokenSymbol,
+			_maxRewardThreshold
+		);
+		//above  initialize going  to revert on second call, so this is safe
+		tokenUsdOracle = _tokenUsdOracle;
+		incentiveController = _incentiveController;
+		aaveUSDOracle = _aaveUSDOracle;
+
+		_approveTokens();
+	}
+
+	/**
+	 * @dev stake some Token
+	 * @param _amount of Token to stake
+	 */
+	function mintInterestToken(uint256 _amount) internal override {
+		lendingPool.deposit(address(token), _amount, address(this), 0);
+	}
+
+	/**
+	 * @dev redeem Token from aave
+	 * @param _amount of token to redeem in Token
+	 */
+	function redeem(uint256 _amount) internal override {
+		lendingPool.withdraw(address(token), _amount, address(this));
+	}
+
+	/**
+	 * @dev Function to redeem aToken for DAI, so reserve knows how to handle it. (reserve can handle dai or cdai)
+	 * @dev _amount of token in iToken
+	 * @return return address of the DAI and amount of the DAI
+	 */
+	function redeemUnderlyingToDAI(uint256 _amount)
+		internal
+		override
+		returns (address, uint256)
+	{
+		lendingPool.withdraw(address(token), _amount, address(this));
+		uint256 redeemedAmount = token.balanceOf(address(this));
+		address[] memory tokenAddress = new address[](1);
+		tokenAddress[0] = address(token);
+		address daiAddress = nameService.getAddress("DAI");
+		uint256 aaveBalance = incentiveController.getRewardsBalance(
+						tokenAddress,
+						address(this));
+		Uniswap uniswapContract =
+			Uniswap(nameService.getAddress("UNISWAP_ROUTER"));
+		address reserveAddress = nameService.getAddress("RESERVE");
+		address[] memory path = new address[](2);
+		path[1] = daiAddress;
+		uint256[] memory swap;
+		uint256 daiFromAave;
+		if(aaveBalance > 0){
+			incentiveController.claimRewards(tokenAddress, aaveBalance, address(this));
```

Comment:
this will not work. 
1. claimrewards gives you stkaave not aave
2. to convert to aave you need cooldown
3. then you can perform redeem

lets just remove stkaave support. just perform "claimRewards(..., recipient)" where recipient is probably the reserve. redeeumunderlying should be refactored to recieve "recipient"

Reply:
changed it